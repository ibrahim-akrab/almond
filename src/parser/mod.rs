//! Parsing for JS
//! # Whitespace Handling
//! All functions named `parse_*` should handle leading whitespace. Preceding whitespace is only handled in top level parse function.

mod expression;
mod functions;
mod identifier;
mod keyword;
mod literal;
mod precedence;
mod statement;
mod util;
pub use expression::*;
pub use functions::*;
pub use identifier::*;
pub use keyword::*;
pub use literal::*;
pub use precedence::*;
pub use statement::*;
pub(crate) use util::*;

use nom::{character::complete::*, multi::*, sequence::*, IResult};

pub fn parse_program(i: &str) -> IResult<&str, &str> {
    nom::bytes::complete::tag("hello")(i)
}
