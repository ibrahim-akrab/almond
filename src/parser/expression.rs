//! Parsing for JS expressions.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{branch::alt, bytes::complete::*, combinator::*, IResult};
use nom_locate::position;

/// Alias for `parse_expr_bp(s, 0)`.
pub fn parse_expr(s: Span) -> IResult<Span, Node> {
    parse_expr_bp(s, 0)
}

/// Parse an atomic expression â€” either a single token that is an
// expression, an expression started by a keyword like `function` or
// `new`, or an expression wrapped in punctuation like `()`, `[]`,
// or `{}`.
pub fn parse_primary_expr(s: Span) -> ParseResult<Node> {
    alt((
        parse_this_expr,
        parse_identifier,
        literal::parse_literal,
        parse_paren_expr,
    ))(s)
}

pub fn parse_this_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, _) = tag("this")(s)?;
    let (s, end) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((s, NodeKind::ThisExpression.with_pos(start, end)))
}

pub fn parse_paren_expr(s: Span) -> ParseResult<Node> {
    delimited(ws0(char('(')), parse_expr, ws0(char(')')))(s)
}

fn parse_opt_expr_in_list(s: Span) -> ParseResult<Option<Node>> {
    alt((
        value(None, peek(char(','))),
        map(parse_expr, |expr| Some(expr)),
    ))(s)
}

pub fn parse_expr_list_with_opt_expr(s: Span) -> ParseResult<Vec<Option<Node>>> {
    terminated(
        separated_list0(ws0(char(',')), parse_opt_expr_in_list),
        // eat trailing comma
        ws0(opt(char(','))),
    )(s)
}

pub fn parse_expr_list(s: Span) -> ParseResult<Vec<Node>> {
    terminated(
        separated_list0(ws0(char(',')), parse_expr),
        // eat trailing comma
        ws0(opt(char(','))),
    )(s)
}

fn parse_computed_member_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (object, property)) = tuple((
        parse_member_expr,
        delimited(ws0(char('[')), parse_expr, char(']')),
    ))(s)?;
    let (s, end): (Span, Span) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((
        s,
        NodeKind::MemberExpression {
            computed: true,
            property: Box::new(property),
            object: Box::new(object),
        }
        .with_pos(start, end),
    ))
}

fn parse_prop_member_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (object, property)) = tuple((
        parse_member_expr,
        preceded(ws0(char('.')), parse_identifier),
    ))(s)?;
    let (s, end): (Span, Span) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((
        s,
        NodeKind::MemberExpression {
            computed: false,
            property: Box::new(property),
            object: Box::new(object),
        }
        .with_pos(start, end),
    ))
}

fn parse_new_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (callee, arguments)) = preceded(
        ws1(keyword_new),
        pair(
            parse_member_expr,
            delimited(ws0(char('(')), parse_expr_list, char(')')),
        ),
    )(s)?;
    let (s, end) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((
        s,
        NodeKind::NewExpression {
            callee: Box::new(callee),
            arguments,
        }
        .with_pos(start, end),
    ))
}

pub fn parse_member_expr(s: Span) -> ParseResult<Node> {
    alt((
        // terminated(parse_primary_expr, not(alt((char('.'), char('['))))),
        parse_prop_member_expr,     // '.'
        parse_computed_member_expr, // '['
        parse_new_expr,             // new
        parse_primary_expr,
    ))(s)
}

pub fn parse_call_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (callee, arguments)) = pair(
        parse_member_expr,
        delimited(ws0(char('(')), parse_expr_list, char(')')),
    )(s)?;
    let (s, end) = position(s)?;
    let (s, _) = sp0(s)?;
    Ok((
        s,
        NodeKind::CallExpression {
            callee: Box::new(callee),
            arguments,
        }
        .with_pos(start, end),
    ))
}

pub fn parse_lhs_expr(s: Span) -> ParseResult<Node> {
    alt((parse_call_expr, parse_member_expr))(s)
}

// fn prefix_expr(s: Span) -> ParseResult<Node> {

// }

// fn postfix_expr(s: Span) -> ParseResult<Node> {

// }

// pub fn parse_unary_expr(s: Span) -> ParseResult<Node> {
//     ws0(alt((prefix_expr, postfix_expr)))(s)
// }

/// Pratt parsing for prefix operators. Called in `parse_expr_bp`.
fn parse_prefix_expr(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, (unary_op, BindingPower(_, right_bp))) = parse_prefix_operator(s)?;
    let (s, rhs) = parse_expr_bp(s, right_bp)?;
    let (s, end) = position(s)?;

    let node_kind = match unary_op {
        PrefixOperator::Unary(unary_op) => NodeKind::UnaryExpression {
            argument: Box::new(rhs),
            operator: unary_op,
            prefix: true,
        },
        PrefixOperator::Update(unary_op) => NodeKind::UpdateExpression {
            argument: Box::new(rhs),
            operator: unary_op,
            prefix: true,
        },
    };
    Ok((s, node_kind.with_pos(start, end)))
}

/// Pratt parsing for expressions with operator precedence.
/// Check out [https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html](https://matklad.github.io/2020/04/13/simple-but-powerful-pratt-parsing.html) to see how Pratt parsing works.
pub fn parse_expr_bp(s: Span, min_bp: i32) -> ParseResult<Node> {
    let (mut s, mut lhs) = alt((parse_prefix_expr, parse_primary_expr))(s)?;

    loop {
        // do not override s just yet
        let (s_tmp, (op, BindingPower(left_bp, right_bp))) = match parse_infix_operator(s) {
            Ok(res) => res,
            Err(_) => break, // do not return from function, just break from loop.
        };

        if left_bp < min_bp {
            break;
        }

        // ok, now we can override s
        s = s_tmp;

        let (s_tmp, rhs) = parse_expr_bp(s, right_bp)?;
        s = s_tmp;

        let start = lhs.clone().start;
        let end = rhs.clone().end;

        let node_kind = match op {
            InfixOperator::Binary(op) => NodeKind::BinaryExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::Logical(op) => NodeKind::LogicalExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::Assignment(op) => NodeKind::AssignmentExpression {
                left: Box::new(lhs),
                right: Box::new(rhs),
                operator: op,
            },
            InfixOperator::DotOperator => NodeKind::MemberExpression {
                object: Box::new(lhs),
                property: Box::new(rhs),
                computed: false,
            },
        };
        lhs = node_kind.with_pos(start, end);
    }

    Ok((s, lhs))
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn smoke_test_this_expr() {
        parse_this_expr("this".into()).unwrap();
        parse_this_expr("notthis".into()).unwrap_err();
    }

    #[test]
    fn smoke_test_primary_expr() {
        parse_primary_expr("this".into()).unwrap();
        parse_primary_expr("myVar".into()).unwrap();
    }

    #[test]
    fn smoke_test_paren_expr() {
        parse_paren_expr("(true)".into()).unwrap();
    }

    #[test]
    #[ignore]
    fn test_member_expr() {
        assert_json_snapshot!(parse_member_expr("a.b".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("a.b.c".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("a[1]".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("a[0]".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("a[[]]".into()).unwrap().1);
    }

    #[test]
    #[ignore]
    fn test_new_expr() {
        assert_json_snapshot!(parse_member_expr("new Array()".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("new Array(1)".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("new Array(1,)".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("new Array(1,2)".into()).unwrap().1);
        assert_json_snapshot!(parse_member_expr("new Foo.Bar(true)".into()).unwrap().1);
    }

    #[test]
    #[ignore]
    fn test_call_expr() {
        assert_json_snapshot!(parse_call_expr("foo()".into()).unwrap().1);
        assert_json_snapshot!(parse_call_expr("foo.bar()".into()).unwrap().1);
        assert_json_snapshot!(parse_call_expr("foo.bar.baz()".into()).unwrap().1);
        assert_json_snapshot!(parse_call_expr("foo.bar(baz)".into()).unwrap().1);
        assert_json_snapshot!(parse_call_expr("foo.bar(baz, 1, 2, 3)".into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_infix() {
        assert_json_snapshot!(parse_expr("1 + 2".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 - 2 - 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 * 2 + 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 + 2 * 3".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 * 2 + 3 * 4".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("(1 + 2) * 3".into()).unwrap().1);
    }
    #[test]
    fn test_expr_bp_prefix() {
        assert_json_snapshot!(parse_expr("-1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("+1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("+(+1)".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("1 + -2".into()).unwrap().1); // same as 1 + (-1)

        assert_json_snapshot!(parse_expr("++x".into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_assignment() {
        assert_json_snapshot!(parse_expr("x = 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x = 1 + 2".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += y += 1".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x += x * x".into()).unwrap().1);
    }

    #[test]
    fn test_expr_bp_member_expr() {
        assert_json_snapshot!(parse_expr("x.y".into()).unwrap().1);
        assert_json_snapshot!(parse_expr("x.y.z".into()).unwrap().1);
    }
}
