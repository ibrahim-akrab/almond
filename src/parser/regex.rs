//! Parsing for JS regexp literals

use crate::ast::*;
use crate::parser::*;

fn regex_body(s: Span) -> ParseResult<String> {
    let escaped_regex_body = preceded(char('\\'), alt((char('\\'), char('/'))));
    fold_many0(
        alt((escaped_regex_body, none_of("/"))),
        format!(""),
        |mut body: String, regex_char| {
            body.push(regex_char);
            body
        },
    )(s)
}

fn regex_flags(s: Span) -> ParseResult<String> {
    fold_many0(
        identifier_continue,
        format!(""),
        |mut flag: String, flag_char| {
            flag.push(flag_char);
            flag
        },
    )(s)
}

pub fn regex_lit(s: Span) -> ParseResult<Node> {
    map(
        spanned(recognize(tuple((
            tag("/"),
            regex_body,
            tag("/"),
            regex_flags,
        )))),
        |(regex_str, start, end)| {
            LiteralValue::RegExp(regex_str.to_string())
                .into_node_kind()
                .with_pos(start, end)
        },
    )(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_regex_lit() {
        assert_json_snapshot!(parse_literal("/abc/g".into()).unwrap().1);
        assert_json_snapshot!(parse_literal(r#"/abc\\/g"#.into()).unwrap().1);
        assert_json_snapshot!(parse_literal(r#"/\\\//g"#.into()).unwrap().1);
    }
}
