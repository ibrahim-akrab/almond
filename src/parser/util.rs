//! Utilities

use crate::parser::*;
use nom::{branch::*, bytes::complete::*, combinator::*, IResult, Parser};

pub type Span<'a> = nom_locate::LocatedSpan<&'a str>;
pub type ParseResult<'a, T> = IResult<Span<'a>, T>;

fn is_whitespace(c: &char) -> bool {
    match c {
        '\t'
        // vertical tab
        | '\x0b'
        // form feed
        | '\x0c'
        | ' '
        // no break space
        | '\u{00a0}'
        // byte order mark
        | '\u{feff}'
        // unicode space seperators
        | '\u{2000}'..='\u{200a}' | '\u{3000}' => true,
        _ => false
    }
}

/// Parses one character if it is whitespace (not including newline and comments).
pub fn whitespace(s: Span) -> ParseResult<()> {
    value((), verify(anychar, is_whitespace))(s)
}

/// # Grammar
/// ```ebnf
/// lineTerminator = "\n" | "\r" | "\u2028" | "\u2029"
/// ```
pub fn line_terminator(s: Span) -> IResult<Span, ()> {
    let (s, _) = one_of("\n\r\u{2028}\u{2029}")(s)?;
    Ok((s, ()))
}

pub fn is_line_terminator(c: char) -> bool {
    match c {
        '\n' | '\r' | '\u{2028}' | '\u{2029}' => true,
        _ => false,
    }
}

/// # Grammar
/// ```ebnf
/// lineTerminatorSequence = "\n" | "\r" ~"\n" | "\u2028" | "\u2029" | "\r\n"
/// ```
pub fn line_terminator_sequence(s: Span) -> IResult<Span, ()> {
    fn parse_carriage_return(s: Span) -> IResult<Span, ()> {
        let (s, _) = tag("\r\n")(s)?;
        Ok((s, ()))
    }
    alt((parse_carriage_return, line_terminator))(s)?;
    Ok((s, ()))
}

fn single_line_comment(s: Span) -> IResult<Span, Span> {
    let (s, _) = tag("//")(s)?;
    let (s, comment) = take_while(|c| !is_line_terminator(c))(s)?;
    let (s, _) = alt((value((), eof), line_terminator))(s)?;
    Ok((s, comment))
}

fn multi_line_comment(s: Span) -> IResult<Span, Span> {
    delimited(tag("/*"), take_until("*/"), tag("*/"))(s)
}

pub fn comment(s: Span) -> IResult<Span, ()> {
    value((), alt((single_line_comment, multi_line_comment)))(s)
}

pub fn sp(s: Span) -> ParseResult<()> {
    alt((whitespace, line_terminator, comment))(s)
}

pub fn sp0(s: Span) -> IResult<Span, ()> {
    if eof::<Span, ()>(s).is_ok() {
        return Ok((s, ()));
    }
    value((), many0(sp))(s)
}

pub fn sp1(s: Span) -> IResult<Span, ()> {
    if eof::<Span, ()>(s).is_ok() {
        return Ok((s, ()));
    }
    value((), many1(sp))(s)
}

/// Alias for `terminated(f, sp0)`.
pub fn ws0<'a, O1, F>(mut f: F) -> impl FnMut(Span<'a>) -> ParseResult<O1>
where
    F: Parser<Span<'a>, O1, nom::error::Error<Span<'a>>>,
{
    move |input: Span<'a>| {
        let (input, o1) = f.parse(input)?;
        sp0.parse(input).map(|(i, _)| (i, o1))
    }
}

/// Alias for `terminated(f, sp1)`.
pub fn ws1<'a, O1, F>(mut f: F) -> impl FnMut(Span<'a>) -> ParseResult<O1>
where
    F: Parser<Span<'a>, O1, nom::error::Error<Span<'a>>>,
{
    move |input: Span<'a>| {
        let (input, o1) = f.parse(input)?;
        sp1.parse(input).map(|(i, _)| (i, o1))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn smoke_test_comment() {
        all_consuming(comment)("// abc".into()).unwrap();
        all_consuming(comment)("/* abc */".into()).unwrap();
        all_consuming(comment)("/* abc\n123 */".into()).unwrap();
        all_consuming(comment)("/* abc\n123 */ foo".into()).unwrap_err();
    }

    #[test]
    fn test_comment() {
        assert_eq!(*comment("// abc\ndef".into()).unwrap().0.fragment(), "def");
        assert_eq!(*comment("// abc\rdef".into()).unwrap().0.fragment(), "def");
        assert_eq!(
            *comment("// abc\u{2028}def".into()).unwrap().0.fragment(),
            "def"
        );
        assert_eq!(
            *comment("// abc\u{2029}def".into()).unwrap().0.fragment(),
            "def"
        );
        assert_eq!(
            *comment("// abc\r\ndef".into()).unwrap().0.fragment(),
            "\ndef"
        );
    }
}
