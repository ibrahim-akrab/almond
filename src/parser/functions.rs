//! Parsing for JS functions and programs.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{branch::alt, bytes::complete::*, combinator::*};

/// Parses a complete JS program
pub fn parse_program(_s: Span) -> ParseResult<Node> {
    todo!();
}

pub fn parse_declaration(s: Span) -> ParseResult<Node> {
    parse_function_declaration(s)
}

pub fn parse_function_declaration(s: Span) -> ParseResult<Node> {
    let parse_function_declaration_signature = pair(
        preceded(ws0(keyword_function), parse_identifier),
        delimited(ws0(tag("(")), parse_formal_param_list, ws0(tag(")"))),
    );

    map(
        spanned(pair(
            parse_function_declaration_signature,
            parse_function_body,
        )),
        |(((id, params), body), start, end)| {
            NodeKind::FunctionDeclaration {
                function: Function {
                    id: Box::new(Some(id)),
                    params,
                    body: Box::new(body),
                },
            }
            .with_pos(start, end)
        },
    )(s)
}

pub fn parse_function_expr(_s: Span) -> ParseResult<Node> {
    todo!();
}

pub fn parse_function_body(s: Span) -> ParseResult<Node> {
    map(
        spanned(delimited(
            ws0(tag("{")),
            parse_function_body_inner,
            ws0(tag("}")),
        )),
        |(body, start, end)| NodeKind::BlockStatement { body }.with_pos(start, end),
    )(s)
}

pub fn parse_function_body_inner(s: Span) -> ParseResult<Vec<Node>> {
    let parse_directive_list = many0(parse_directive);

    map(
        pair(parse_directive_list, parse_stmt_list),
        |(mut directives, stmts)| {
            directives.extend(stmts);
            directives
        },
    )(s)
}

pub fn parse_formal_param_list(s: Span) -> ParseResult<Vec<Node>> {
    terminated(
        separated_list0(ws0(char(',')), parse_formal_param),
        // trailing comma
        opt(ws0(char(','))),
    )(s)
}

pub fn parse_formal_param(s: Span) -> ParseResult<Node> {
    parse_identifier(s)
}

/// Example: `"use strict";`
pub fn parse_directive(s: Span) -> ParseResult<Node> {
    let parse_directive_literal = spanned(terminated(ws0(string_lit), opt(ws0(semi))));
    map(parse_directive_literal, |(node, start, end)| {
        match &node.kind {
            NodeKind::Literal { value } => match value {
                LiteralValue::String(directive) => {
                    let directive = directive.clone();
                    NodeKind::ExpressionStatement {
                        expression: Box::new(node),
                        directive: Some(directive),
                    }
                    .with_pos(start, end)
                }
                _ => unreachable!(
                    "string_lit should return a literal with value LiteralValue::String"
                ),
            },
            _ => unreachable!("string_lit should return a Node::Identifier"),
        }
    })(s)
}

pub fn parse_source_elem(s: Span) -> ParseResult<Node> {
    alt((parse_declaration, parse_stmt))(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_directive() {
        assert_json_snapshot!(parse_directive("\"use strict\"".into()).unwrap().1);
    }

    #[test]
    fn test_function_declaration() {
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    "use strict"
                    return 1
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a() {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a(param1, param2) {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a(param1, param2,) {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    "use strict"
                    "also directive"
                    a;
                    "not directive"
                    return 1
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }
}
