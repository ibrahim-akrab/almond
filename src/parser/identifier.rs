use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{
    branch::alt, bytes::complete::*, character::complete::*, character::*, combinator::*,
    number::complete::*, IResult,
};
use nom_locate::position;

/// Parses any 4 hex digits unicode escape sequence
fn parse_unicode_escape_sequence(s: Span) -> IResult<Span, char> {
    let (s, _) = tag("\\u")(s)?;
    let (s, hex_str) = recognize(count(one_of("1234567890abcdefABCDEF"), 4))(s)?;
    let hex_u32 = u32::from_str_radix(*hex_str, 16).unwrap(); // FIXME
    let char = std::char::from_u32(hex_u32).unwrap(); // FIXME
    Ok((s, char))
}

pub fn parse_identifier(s: Span) -> IResult<Span, Node> {
    fn parse_identifier_start(s: Span) -> IResult<Span, char> {
        let (s, c) = verify(
            alt((parse_unicode_escape_sequence, none_of(""))),
            |c: &char| match c {
                c if unicode_xid::UnicodeXID::is_xid_start(*c) => true,
                '$' | '_' => true,
                _ => false,
            },
        )(s)?;
        Ok((s, c))
    }
    fn parse_identifier_continue(s: Span) -> IResult<Span, char> {
        let (s, c) = verify(
            alt((parse_unicode_escape_sequence, none_of(""))),
            |c: &char| match c {
                c if unicode_xid::UnicodeXID::is_xid_continue(*c) => true,
                '$' | '_' => true,
                _ => false,
            },
        )(s)?;
        Ok((s, c))
    }

    let (s, start) = position(s)?;
    let (s, name) = map(
        pair(parse_identifier_start, many0(parse_identifier_continue)),
        |(c, v)| format!("{}{}", c, v.into_iter().collect::<String>()),
    )(s)?;
    let (s, end) = position(s)?;

    Ok((s, NodeKind::Identifier { name }.with_pos(start, end)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_identifier() {
        all_consuming(parse_identifier)("myVar".into()).unwrap();
        all_consuming(parse_identifier)("abc123".into()).unwrap();
        all_consuming(parse_identifier)("$".into()).unwrap();
        all_consuming(parse_identifier)("_".into()).unwrap();
        all_consuming(parse_identifier)("$elem".into()).unwrap();
        all_consuming(parse_identifier)("_elem".into()).unwrap();
        all_consuming(parse_identifier)("$123".into()).unwrap();
        all_consuming(parse_identifier)("_123".into()).unwrap();

        // bad identifier
        all_consuming(parse_identifier)("123abc".into()).unwrap_err();
    }

    #[test]
    fn test_identifier_with_unicode_escape() {
        assert_eq!(
            parse_identifier(r#"abc\u0061"#.into()).unwrap().1.kind,
            NodeKind::Identifier {
                name: "abca".into()
            }
        );
        assert_eq!(
            parse_identifier(r#"abc\u0061\u0062123"#.into()).unwrap().1.kind,
            NodeKind::Identifier {
                name: "abcab123".into()
            }
        );
    }
}
