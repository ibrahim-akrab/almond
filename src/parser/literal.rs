//! Parsing for JS literals.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{branch::alt, bytes::complete::*, combinator::*, number::complete::*, IResult};
use nom_locate::position;

pub fn parse_literal(s: Span) -> IResult<Span, Node> {
    ws0(alt((null_lit, bool_lit, array_lit)))(s)
}

pub fn null_lit(s: Span) -> IResult<Span, Node> {
    let (s, start) = position(s)?;
    let (s, _) = tag("null")(s)?;
    let (s, end) = position(s)?;
    Ok((s, LiteralValue::Null.into_node_kind().with_pos(start, end)))
}

pub fn bool_lit(s: Span) -> IResult<Span, Node> {
    let (s, start) = position(s)?;
    let (s, val) = alt((
        value(false, pair(tag("false"), not(identifier_continue))),
        value(true, pair(tag("true"), not(identifier_continue))),
    ))(s)?;
    let (s, end) = position(s)?;
    Ok((
        s,
        LiteralValue::Boolean(val)
            .into_node_kind()
            .with_pos(start, end),
    ))
}

pub fn numeric_lit(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;

    let (s, sign) =
        ws0(opt(char('-')))(s).map(|(s, minus)| (s, if minus.is_some() { -1 } else { 1 }))?;

    let (s, num) =
        alt((octal_int_lit, hex_int_lit, decimal_lit))(s).map(|(s, num)| (s, sign as f64 * num))?;
    let (s, end) = position(s)?;
    Ok((
        s,
        LiteralValue::Number(num)
            .into_node_kind()
            .with_pos(start, end),
    ))
}

fn decimal_lit(s: Span) -> ParseResult<f64> {
    recognize_float(s).map(|(s, float_str)| (s, float_str.parse::<f64>().unwrap()))
}

fn octal_int_lit(s: Span) -> ParseResult<f64> {
    let (s, octal_str) = preceded(char('0'), oct_digit1)(s)?;
    Ok((
        s,
        i64::from_str_radix(octal_str.fragment(), 8).unwrap() as f64,
    ))
}

fn hex_int_lit(s: Span) -> ParseResult<f64> {
    let (s, hex_str) = preceded(alt((tag("0x"), tag("0X"))), hex_digit1)(s)?;
    Ok((
        s,
        i64::from_str_radix(hex_str.fragment(), 16).unwrap() as f64,
    ))
}

pub fn array_lit(s: Span) -> IResult<Span, Node> {
    let (s, start) = position(s)?;
    let (s, expr_list) = delimited(terminated(char('['), spaces0), parse_expr_list, char(']'))(s)?;
    let (s, end) = position(s)?;
    spaces0(s)?;
    Ok((
        s,
        NodeKind::ArrayExpression {
            elements: expr_list,
        }
        .with_pos(start, end),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bool_lit() {
        bool_lit("true".into()).unwrap();
        bool_lit("false".into()).unwrap();
    }

    #[test]
    fn test_bool_lit_with_trailing_whitespace() {
        bool_lit("true ".into()).unwrap();
        bool_lit("false ".into()).unwrap();
    }

    #[test]
    fn test_bad_bool_lit() {
        bool_lit("bad".into()).unwrap_err();
        bool_lit("badfalse".into()).unwrap_err();
        bool_lit("falsebad".into()).unwrap_err();
    }

    #[test]
    fn test_numeric_lit() {
        assert_eq!(
            numeric_lit("123".into()).unwrap().1.kind,
            LiteralValue::Number(123.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("0123".into()).unwrap().1.kind,
            LiteralValue::Number(83.0).into_node_kind() // in octal
        );
        assert_eq!(
            numeric_lit("0x123".into()).unwrap().1.kind,
            LiteralValue::Number(291.0).into_node_kind() // in hex
        );
        assert_eq!(
            numeric_lit("0".into()).unwrap().1.kind,
            LiteralValue::Number(0.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("-123".into()).unwrap().1.kind,
            LiteralValue::Number(-123.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("-0".into()).unwrap().1.kind,
            LiteralValue::Number(-0.0).into_node_kind()
        );
        assert_eq!(
            numeric_lit("-0123".into()).unwrap().1.kind,
            LiteralValue::Number(-83.0).into_node_kind() // in octal
        );
        assert_eq!(
            numeric_lit("-0x123".into()).unwrap().1.kind,
            LiteralValue::Number(-291.0).into_node_kind() // in octal
        );
        assert_eq!(
            numeric_lit("- 123".into()).unwrap().1.kind,
            LiteralValue::Number(-123.0).into_node_kind()
        );
    }

    #[test]
    fn test_array_lit() {
        array_lit("[]".into()).unwrap();
        array_lit("[true, true]".into()).unwrap();
    }

    #[test]
    fn test_array_lit_with_whitespace() {
        array_lit("[ true, true ]\n".into()).unwrap();
        array_lit("[   true,\n true\t ]\t".into()).unwrap();
    }

    #[test]
    fn test_array_lit_with_trailing_comma() {
        array_lit("[ true, true, ]\n".into()).unwrap();
        array_lit("[   true,\n true,\t ]\t".into()).unwrap();
    }
}
