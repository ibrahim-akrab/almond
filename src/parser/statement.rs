//! Parsing for JS statements.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::{branch::alt, bytes::complete::*, combinator::*};
use nom_locate::position;

pub fn parse_stmt(s: Span) -> ParseResult<Node> {
    alt((parse_block, parse_var_stmt, parse_empty_stmt))(s)
}

pub fn parse_block(s: Span) -> ParseResult<Node> {
    map(
        spanned(delimited(ws0(tag("{")), parse_stmt_list, tag("}"))),
        |(body, start, end)| NodeKind::BlockStatement { body }.with_pos(start, end),
    )(s)
}

pub fn parse_stmt_list(s: Span) -> ParseResult<Vec<Node>> {
    many0(parse_stmt)(s)
}

pub fn parse_var_stmt(s: Span) -> ParseResult<Node> {
    let initializer = preceded(ws0(tag("=")), parse_expr_no_seq);
    let var_declaration = map(
        tuple((position, pair(parse_identifier, opt(initializer)), position)),
        |(start, (id, init), end)| {
            NodeKind::VariableDeclarator {
                id: Box::new(id),
                init: Box::new(init),
            }
            .with_pos(start, end)
        },
    );
    let parse_declaration_list = separated_list1(ws0(tag(",")), ws0(var_declaration));

    let (s, start) = position(s)?;

    let (s, (declarations, end)) = spanned_end(delimited(
        ws1(keyword_var),
        parse_declaration_list,
        opt(semi),
    ))(s)?;

    Ok((
        s,
        NodeKind::VariableDeclaration {
            declarations,
            kind: VariableDeclarationKind::Var,
        }
        .with_pos(start, end),
    ))
}

pub fn parse_empty_stmt(s: Span) -> ParseResult<Node> {
    map(spanned(tag(";")), |(_, start, end)| {
        NodeKind::EmptyStatement.with_pos(start, end)
    })(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_var_stmt() {
        assert_json_snapshot!(parse_stmt("var x = 1;".into()).unwrap().1);
        assert_json_snapshot!(parse_stmt("var x, y = 1;".into()).unwrap().1);
        assert_json_snapshot!(parse_stmt("var x = 1, y;".into()).unwrap().1);
        assert_json_snapshot!(parse_stmt("var x".into()).unwrap().1); // auto insert semi
        assert_json_snapshot!(parse_stmt("var x\t".into()).unwrap().1);
    }

    #[test]
    fn test_empty_stmt() {
        assert_json_snapshot!(parse_stmt(";".into()).unwrap().1);
        assert_json_snapshot!(parse_stmt(";\t".into()).unwrap().1);
    }
}
