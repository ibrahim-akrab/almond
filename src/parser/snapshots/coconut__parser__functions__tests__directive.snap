---
source: src/parser/functions.rs
expression: "parse_directive(\"\\\"use strict\\\"\".into()).unwrap().1"
---
{
  "type": "ExpressionStatement",
  "expression": {
    "type": "Literal",
    "value": "use strict",
    "start": 0,
    "end": 12
  },
  "directive": "use strict",
  "start": 0,
  "end": 12
}
